BEST_PRACTICE.md

```diff
@@ -0,0 +1,247 @@
+│   │
+│   ├── commander/
+│   │   ├─ commander.h
+│   │   ├─ commander.c
+│   │   └─ package.json
+│   │
+│   └── logger/
+│       ├── logger.h
+│       ├── logger.c
+│       └── package.json
+│
+├── LICENSE
+│
+├── Makefile
+│
+├── README.md
+│
+├── package.json
+│
+├── src/
+│   ├── main.c
+│   ├── function.c
+│   ├── function.h
+│   └── etc...
+│
+└── test.sh
+```
+
+`test.sh` is your test script. `src` is the directory where you source
+code is in. And `deps` is the directory where you libraries will be
+downloaded when you run `clib install <username/library>`.
+
+Knowing all of that, lets have a look at an example Makefile.
+
+### Example Makefile
+
+```makefile
+# your c compiler
+CC = gcc
+
+# where to install
+PREFIX = /usr/local/bin
+
+# your output file
+TARGET = project-name
+
+# your project name
+TARGET = your-project
+
+CFLAGS = -Ideps -Wall
+
+# all the source files
+SRC = $(wildcard src/*.c)
+SRC += $(wildcard deps/*/*.c)
+
+OBJS = $(SRC:.c=.o)
+
+.PHONY:
+all: $(TARGET)
+
+.PHONY:
+$(TARGET): $(OBJS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -o $(TARGET) $(OBJS)
+
+.PHONY:
+%.o: %.c
+	$(CC) $(DEP_FLAG) $(CFLAGS) $(LDFLAGS) -o $@ -c $<
+
+.PHONY:
+clean:
+	rm -f $(OBJS)
+
+.PHONY:
+install: $(TARGET)
+	cp -f $(TARGET) $(PREFIX)
+
+.PHONY:
+uninstall: $(PREFIX)/$(TARGET)
+	rm -f $(PREFIX)/$(TARGET)
+```
+
+This is a basic Makefile, and should work for most of your projects.
+
+You *could* have your Makefile install the libraries upon running it, but you
+would only need to do that to get the latest version of the library(s), in this
+case you proboly don't want that. You typicly want yo get the lates stable version
+for that library. By having a `package.json` file in your project repo, you can
+specify what packages you need, and what version of that package. Now have a look
+at a example `package.json` file for your project: (executable package)
+
+### Example package.json for exacutable project
+
+```
+{
+  "name": "executable-name",
+  "version": "1.0.0",
+  "repo": "your-github-name/project-name",
+  "dependencies": {
+    "stephenmathieson/trim.c": "0.0.2",
+    "clibs/commander": "1.3.2",
+    "clibs/logger": "0.0.1",
+  },
+  "install": "make install",
+  "uninstall": "make uninstall"
+}
+```
+
+Starting from the top, `"name"` is your package name. `"version"` is your package version.
+`"repo"` is the location of your project, (not including the `https://github.com/`).
+`"dependencies"` is all the dependencies your project requiers, along with there version.
+`"install"` is the command to install your program (ran as root), (tip: if your project
+requiers more then one command to install it, like need to run `./configure`, before `make`,
+then do this: `"install": "./configure && make && make install"`). `"uninstall"` is the command
+to uninstall your project, [more on that later](#install-and-uninstall-executables).
+
+***NOTE:** Make sure you have a relese as the same version in your `package.json` file, otherwise
```

Comment:
This doesn't look like valid Markdown. Why are there so many `*`s here?

Reply:
I'll try to fix that...